x-pack/plugin/enrich/src/test/java/org/elasticsearch/xpack/enrich/EnrichPolicyRunnerTests.java

```diff
@@ -212,46 +181,390 @@ public void testRunnerMultiSource() throws Exception {
+                            .endObject()
+                        .endObject()
+                    .endObject()
+                .endObject()
+            .endObject()
+            .endObject();
+        CreateIndexResponse createResponse = client().admin().indices().create(new CreateIndexRequest(sourceIndex)
+            .mapping(MapperService.SINGLE_MAPPING_NAME, mappingBuilder)).actionGet();
+        assertTrue(createResponse.isAcknowledged());
 
-        EnrichPolicyRunner enrichPolicyRunner = new EnrichPolicyRunner(policyName, policy, listener, clusterService,
-            client(), resolver, () -> createTime, randomIntBetween(1, 10000));
+        String policyName = "test1";
+        List<String> enrichFields = List.of("nesting.field2", "missingField");
+        EnrichPolicy policy = new EnrichPolicy(EnrichPolicy.EXACT_MATCH_TYPE, null, List.of(sourceIndex), "key", enrichFields);
+
+        final long createTime = randomNonNegativeLong();
+        final AtomicReference<Exception> exception = new AtomicReference<>();
+        final CountDownLatch latch = new CountDownLatch(1);
+        ActionListener<PolicyExecutionResult> listener = createTestListener(latch, exception::set);
+        EnrichPolicyRunner enrichPolicyRunner = createPolicyRunner(policyName, policy, listener, createTime);
 
         logger.info("Starting policy run");
+        enrichPolicyRunner.run();
+        latch.await();
+        if (exception.get() != null) {
+            Exception thrown = exception.get();
+            assertThat(thrown, instanceOf(ElasticsearchException.class));
+            assertThat(thrown.getMessage(), containsString("Enrich policy execution for [" + policyName +
+                "] failed while validating field mappings for index [" + sourceIndex + "]"));
+            assertThat(thrown.getCause().getMessage(), containsString("Could not traverse mapping to field [nesting.field2]. " +
+                "The [nesting] field must be regular object but was [nested]."));
+        } else {
+            fail("Expected exception but nothing was thrown");
+        }
+    }
+
+    public void testRunnerObjectSourceMapping() throws Exception {
+        final String sourceIndex = "source-index";
+        XContentBuilder mappingBuilder = JsonXContent.contentBuilder();
+        mappingBuilder.startObject()
+            .startObject(MapperService.SINGLE_MAPPING_NAME)
+                .startObject("properties")
+                    .startObject("data")
+                        .startObject("properties")
+                            .startObject("field1")
+                                .field("type", "keyword")
+                            .endObject()
+                            .startObject("field2")
+                                .field("type", "integer")
+                            .endObject()
+                            .startObject("field3")
+                                .field("type", "keyword")
+                            .endObject()
```

Comment:
maybe also add a test that has a mapping with two nested object fields (level1.level2.*)?